File: day4/go/p1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var letterValues = map[rune]int{
	'X': 1,
	'M': 2,
	'A': 3,
	'S': 4,
}

func calculateHash(word string) int {
	hash := 0
	for _, char := range word {
		hash = hash*21 + letterValues[char]
	}
	return hash
}

func readMatrix(fileName string) ([][]rune, error) {
	file, err := os.Open(fileName)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var matrix [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		matrix = append(matrix, []rune(line))
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return matrix, nil
}

func isOnBounds(row, col, rows, cols int) bool {
	return row >= 0 && row < rows && col >= 0 && col < cols
}

func findOccurrences(matrix [][]rune, targetWord string, targetHash int) int {
	rows := len(matrix)
	if rows == 0 {
		return 0
	}
	cols := len(matrix[0])
	wordLength := len(targetWord)
	occurrences := 0

	directions := [][2]int{
		{0, 1},
		{1, 0},
		{1, 1},
		{1, -1},
		{-1, 0},
		{0, -1},
		{-1, 1},
		{-1, -1},
	}

	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			for _, dir := range directions {
				dRow, dCol := dir[0], dir[1]
				hash := 0
				valid := true

				for i := 0; i < wordLength; i++ {
					newRow := row + dRow*i
					newCol := col + dCol*i

					if !isOnBounds(newRow, newCol, rows, cols) {
						valid = false
						break
					}

					char := matrix[newRow][newCol]
					hash = hash*21 + letterValues[char]
				}

				if valid && hash == targetHash {
					occurrences++
				}
			}
		}
	}

	return occurrences
}

func main() {
	// Lê o arquivo input.txt
	matrix, err := readMatrix("input.txt")
	if err != nil {
		fmt.Println("Erro ao ler o arquivo:", err)
		return
	}

	// Define a palavra-alvo e calcula o hash
	targetWord := "XMAS"
	targetHash := calculateHash(targetWord)

	// Procura a palavra na matriz
	result := findOccurrences(matrix, targetWord, targetHash)

	fmt.Printf("O resultado é: %d\n", result)
}

File: day4/lua/p1.lua
local letterValues = { X = 1, M = 2, A = 3, S = 4 }

local function calculateHash(word)
	local hash = 0
	for i = 1, #word do
		local char = word:sub(i, i)
		hash = hash * 21 + (letterValues[char] or 0)
	end
	return hash
end

local function isOnBounds(row, col, rows, cols)
	return row >= 1 and row <= rows and col >= 1 and col <= cols
end

local function findOccurrences(matrix, targetWord, targetHash)
	local rows = #matrix
	local cols = #matrix[1]
	local wordLength = #targetWord
	local occurrences = 0

	local directions = {
		{ 0, 1 },
		{ 1, 0 },
		{ 1, 1 },
		{ 1, -1 },
		{ -1, 0 },
		{ 0, -1 },
		{ -1, 1 },
		{ -1, -1 },
	}

	for row = 1, rows do
		for col = 1, cols do
			for _, dir in ipairs(directions) do
				local dRow, dCol = dir[1], dir[2]
				local hash = 0
				local valid = true

				for i = 0, wordLength - 1 do
					local newRow = row + dRow * i
					local newCol = col + dCol * i

					if not isOnBounds(newRow, newCol, rows, cols) then
						valid = false
						break
					end

					local char = matrix[newRow][newCol]
					hash = hash * 21 + (letterValues[char] or 0)
				end

				if valid and hash == targetHash then
					occurrences = occurrences + 1
				end
			end
		end
	end

	return occurrences
end

local function readMatrix(fileName)
	local matrix = {}
	for line in io.lines(fileName) do
		local row = {}
		for char in line:gmatch(".") do
			table.insert(row, char)
		end
		table.insert(matrix, row)
	end
	return matrix
end

local inputFileName = "input.txt"
local targetWord = "XMAS"
local targetHash = calculateHash(targetWord)

local matrix = readMatrix(inputFileName)
local result = findOccurrences(matrix, targetWord, targetHash)

print("The result is", result)
